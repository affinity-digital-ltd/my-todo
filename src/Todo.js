import React, { Component } from 'react';
import TodoItem from './TodoItem'

import './scss/todo.scss'

class Todo extends Component {
  constructor () {
    super()

    this.state = {
      todos: [],
      value: ''
    }
    this.addItemToTodo = this.addItemToTodo.bind(this)
    this.deleteItem = this.deleteItem.bind(this)
    this.handleChange = this.handleChange.bind(this)
  }

  addItemToTodo (event) {
    event.preventDefault() // Do not submit the form and cause the page to refresh
    const { todos } = this.state
    const item = this.state.value

    todos.push(item)

    this.setState({
      todos: todos,
      value: ''
    })
  }

  buildTodoList () {
    const { todos } = this.state

    return todos.map((todo, index) => {
      return <li key={todo}>
        <TodoItem item={todo} deleteItem={() => this.deleteItem(index)} />
      </li>
    })
  }

  handleChange (event) {
    this.setState({
      value: event.target.value
    })
  }

  deleteItem (item) {
    const { todos } = this.state

    todos.splice(item, 1)

    this.setState({
      todos: todos
    })
  }

  render () {
    return (
      <>
        <form onSubmit={this.addItemToTodo} className='c-form'>
          <div className='c-form__input-group'>
            <input type='text' placeholder='What would you like to do?' value={this.state.value} name='todoItem' autoComplete='off' onChange={this.handleChange} />
            <button type='submit' className='c-button'>Add</button>
          </div>
        </form>
        <ul>
          {this.buildTodoList()}
        </ul>
      </>
    )
  }
}

export default Todo;
