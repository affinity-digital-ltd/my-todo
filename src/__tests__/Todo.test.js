import React from 'react'
import Todo from '../Todo'

import { shallow, mount } from 'enzyme'

describe('<Todo />', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(<Todo />)

    expect(wrapper).toMatchSnapshot()
  })

  it('should add a new item to the list', () => {
    const wrapper = mount(<Todo />)

    const input = wrapper.find('[name="todoItem"]')
    input.instance().value = 'test'
    input.simulate('change')
    const button = wrapper.find('[type="submit"]')
    button.simulate('submit')

    expect(wrapper.find('.c-todo-item__content').text()).toEqual('test')
  })

  it('should mark the item as done', () => {
    const wrapper = mount(<Todo />)

    const input = wrapper.find('[name="todoItem"]')
    input.simulate('change', { target: { value: 'test' } })
    const button = wrapper.find('[type="submit"]')
    button.simulate('submit')

    const checkbox = wrapper.find('[type="checkbox"]')
    checkbox.simulate('click', { target: { checked: true } })

    expect(wrapper.find('span').hasClass('u-strikethrough')).toEqual(true)
  })

  it('should remove the todo item', () => {
    const wrapper = mount(<Todo />)

    const input = wrapper.find('[name="todoItem"]')
    input.simulate('change', { target: { value: 'test' } })
    const button = wrapper.find('[type="submit"]')
    button.simulate('submit')

    expect(wrapper.find('span').length).toEqual(1)
    
    const del = wrapper.find('[data-behavior="delete"]')
    del.simulate('click')

    expect(wrapper.find('span').length).toEqual(0)
  })
})
