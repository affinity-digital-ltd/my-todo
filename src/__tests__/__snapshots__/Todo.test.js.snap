// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Todo /> should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Todo />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <form
          className="c-form"
          onSubmit={[Function]}
        >
          <div
            className="c-form__input-group"
          >
            <input
              autoComplete="off"
              name="todoItem"
              onChange={[Function]}
              placeholder="What would you like to do?"
              type="text"
              value=""
            />
            <button
              className="c-button"
              type="submit"
            >
              Add
            </button>
          </div>
        </form>,
        <ul />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="c-form__input-group"
          >
            <input
              autoComplete="off"
              name="todoItem"
              onChange={[Function]}
              placeholder="What would you like to do?"
              type="text"
              value=""
            />
            <button
              className="c-button"
              type="submit"
            >
              Add
            </button>
          </div>,
          "className": "c-form",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                autoComplete="off"
                name="todoItem"
                onChange={[Function]}
                placeholder="What would you like to do?"
                type="text"
                value=""
              />,
              <button
                className="c-button"
                type="submit"
              >
                Add
              </button>,
            ],
            "className": "c-form__input-group",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "autoComplete": "off",
                "name": "todoItem",
                "onChange": [Function],
                "placeholder": "What would you like to do?",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Add",
                "className": "c-button",
                "type": "submit",
              },
              "ref": null,
              "rendered": "Add",
              "type": "button",
            },
          ],
          "type": "div",
        },
        "type": "form",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
        },
        "ref": null,
        "rendered": Array [],
        "type": "ul",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <form
            className="c-form"
            onSubmit={[Function]}
          >
            <div
              className="c-form__input-group"
            >
              <input
                autoComplete="off"
                name="todoItem"
                onChange={[Function]}
                placeholder="What would you like to do?"
                type="text"
                value=""
              />
              <button
                className="c-button"
                type="submit"
              >
                Add
              </button>
            </div>
          </form>,
          <ul />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="c-form__input-group"
            >
              <input
                autoComplete="off"
                name="todoItem"
                onChange={[Function]}
                placeholder="What would you like to do?"
                type="text"
                value=""
              />
              <button
                className="c-button"
                type="submit"
              >
                Add
              </button>
            </div>,
            "className": "c-form",
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  autoComplete="off"
                  name="todoItem"
                  onChange={[Function]}
                  placeholder="What would you like to do?"
                  type="text"
                  value=""
                />,
                <button
                  className="c-button"
                  type="submit"
                >
                  Add
                </button>,
              ],
              "className": "c-form__input-group",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "autoComplete": "off",
                  "name": "todoItem",
                  "onChange": [Function],
                  "placeholder": "What would you like to do?",
                  "type": "text",
                  "value": "",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Add",
                  "className": "c-button",
                  "type": "submit",
                },
                "ref": null,
                "rendered": "Add",
                "type": "button",
              },
            ],
            "type": "div",
          },
          "type": "form",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
          },
          "ref": null,
          "rendered": Array [],
          "type": "ul",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
